src/bytes.rs

```diff
@@ -1028,7 +1028,9 @@ unsafe fn shallow_clone_vec(
     // ordering will synchronize with the `compare_and_swap`
     // that happened in the other thread and the `Shared`
     // pointed to by `actual` will be visible.
-    let actual = atom.compare_and_swap(ptr as _, shared as _, Ordering::AcqRel);
+    let actual = atom
+        .compare_exchange(ptr as _, shared as _, Ordering::AcqRel, Ordering::Acquire)
+        .unwrap_or_else(|x| x);
```

Comment:
I had started on this too, and figured I'd make use of the `Result` and not compare a second time below, in case you want it:

```rust
match atom.compare_exchange(ptr as _, shared _, Ordering::Release, Ordering::Acquire) {
    Ok(actual) => {
        debug_assert!(actual as usize == ptr as usize);
        // The upgrade was successful, the new handle can be
        // returned.
        Bytes {
            ptr: offset,
            len,
            data: AtomicPtr::new(shared as _),
            vtable: &SHARED_VTABLE,
        }
    },
    Err(actual) => {
        // The upgrade failed, a concurrent clone happened. Release
        // the allocation that was made in this thread, it will not
        // be needed.
        let shared = Box::from_raw(shared);
        mem::forget(*shared);

        // Buffer already promoted to shared storage, so increment ref
        // count.
        shallow_clone_arc(actual as _, offset, len)
    }
}
```

Reply:
The first argument should still be `AcqRel`, no?

Reply:
Hm, yea probably. I guess at least to keep the exact code that `compare_and_swap` is doing in libstd.

Reply:
`Ordering::Release, Ordering::Acquire` won't even work, the 2nd ordering needs to be a subset of the first.